### PR TITLE
[RW-3122][risk=low] Convert workspace-wrapper component from Angular to React

### PR DESCRIPTION
**Pull request by ericsong**
Some notes/decisions made
- Hard coded the workspace nav bar tab into the routeConfigData on each route. Previously this was derived from the tree structure of the Angular routes but this no longer exists on the flat structure of the React routes. I could have computed the nav bar tab from the URL but I thought it would be more clear to embed it into the route config.
- There was a bug in data-component.tsx where new resources would not load when the workspace changed. I decided to convert the component to use React Hooks while I was in there. 
- Two React components had their `withRouteData` HOC embedded into the component itself instead of the `app-routing` component. This broke their tests because of how I'm using `withRouteData` to set store values and it's call to `useParams()` was breaking. We can switch this back later once we refactor `withRouteData` but I do think I prefer keeping the routing stuff completely separate from the React component.
- The current implementation of the Angular/React routes forces a full rerender (unmount/remount) of the Reactified portions of the App since Angular will blow away the components and recreate them when the user navigates to a new route. This is inefficient on one hand but also causes some visual wonkiness and forces some less than ideal technical choices.
  - One component where this particularly matters is the help-sidebar component because it was originally designed to be mounted once on application load. Now that it's rendered by React (through Angular router), it's recreated on every new route and this messes up some of the animations as it transitions between pages. 
  - I'm also unable to set up subscriptions on the help-sidebar since they get unsubscribed on every new route when the component is unmounted.

- I spent quite a few hours digging into a tough bug caused rendering the help-sidebar more quickly in the React version of workspace-wrapper.
  - First, I'll describe the state of things before this PR.
    - On page load, we make a call to `getRuntime()` in order to get the status of the runtime and render the status icon in the help sidebar
    - This `getRuntime()` call it made by the... https://github.com/all-of-us/workbench/blob/9cabb3b636bdbdc1f8b9d6cd552bbf7b93db1f2f/ui/src/app/utils/runtime-utils.tsx#L221 `useRuntime()` hook which is loaded by...  https://github.com/all-of-us/workbench/blob/9cabb3b636bdbdc1f8b9d6cd552bbf7b93db1f2f/ui/src/app/utils/runtime-utils.tsx#L401 `withRuntimeStore()` HOC in help-sidebar https://github.com/all-of-us/workbench/blob/9cabb3b636bdbdc1f8b9d6cd552bbf7b93db1f2f/ui/src/app/components/help-sidebar.tsx#L263
    - `useRuntime()` needs to be called with a workspaceNamespace argument and in this flow, it is fetched from the `runtimeStore`.  https://github.com/all-of-us/workbench/blob/9cabb3b636bdbdc1f8b9d6cd552bbf7b93db1f2f/ui/src/app/utils/runtime-utils.tsx#L398
    - `runtimeStore` is updated by a subscription to the `urlParamsStore` which is started in the constructor for `workspace-wrapper/component` https://github.com/all-of-us/workbench/blob/9cabb3b636bdbdc1f8b9d6cd552bbf7b93db1f2f/ui/src/app/pages/workspace/workspace-wrapper/component.ts#L118
    - The `urlParamsStore` subscription event is emitted after an Angular "NavigationEnd" router event https://github.com/all-of-us/workbench/blob/9cabb3b636bdbdc1f8b9d6cd552bbf7b93db1f2f/ui/src/app/pages/app/component.ts#L92
    - The net result is that all these lines of code need to executed in just the right order for the `runtimeStore` to be populated with a valid workspaceNamespace value by the time `useRuntime()` is called.
    - If that does NOT happen, `useRuntime()` will be called with a null workspaceNamespace and every call to `getRuntime()` will fail since it's made with a null argument. https://github.com/all-of-us/workbench/blob/9cabb3b636bdbdc1f8b9d6cd552bbf7b93db1f2f/ui/src/app/utils/runtime-utils.tsx#L221
    - On top the initial `getRuntime()` call failing, there is a runtime poller set up by `workspace-wrapper` which calls `getRuntime()` on a regular interval. These calls will also fail because it's reading a null workspaceNamespace from the `runtimeStore`
    - Even when the all the subscription/event code is run and a valid workspaceNamespace is provided to the `runtimeStore`, it will be overwritten by a null value because of the async error handler in `useRuntime()` which nulls out the `runtimeStore` if the `getRuntime()` call fails. https://github.com/all-of-us/workbench/blob/9cabb3b636bdbdc1f8b9d6cd552bbf7b93db1f2f/ui/src/app/utils/runtime-utils.tsx#L217
    - This probably hints at a larger issue of the difficulty in tracking state changes in our application but I was able to fix this for now by adding a check to `useRuntime()` that exits early if a null workspaceNamespace is given. https://github.com/all-of-us/workbench/pull/5199/files#diff-f53435d20207d854b5a65f4be0494bbeb1390492c2f4a1ac7226b5f90453931dR216
   
---
**PR checklist**

- [ ] This PR meets the Acceptance Criteria in the JIRA story
- [ ] The JIRA story has been moved to Dev Review
- [ ] This PR includes appropriate unit tests
- [ ] I have run and tested this change locally
- [ ] I have run the E2E tests on ths change against my local UI and/or API server with `yarn test-local` or [yarn test-local-devup](https://github.com/all-of-us/workbench/blob/master/e2e/README.md#examples)
- [ ] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer
- [ ] If this includes an API change, I have updated the appropriate Swagger definitions and notified API consumers
- [ ] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later

### CODE REVIEW

File: ui/src/app/app-routing.module.ts [hunk 1]
```diff
@@ -111,10 +110,14 @@ const routes: Routes = [
               {
                 /* TODO The children under ./views need refactoring to use the data
                  * provided by the route rather than double-requesting it.
+                 *
+                 * TODO angular2react: ideally, we should be able to just render the AppComponent here and let
```
Comment (calbach) [inline ui/src/app/app-routing.module.ts:114]: One idea in the interim could be to let the signed in component own `/**`, and map once to AppRouting. I'm not sure whether this will continue to cause issues with angular router interactions.

**Review by calbach — COMMENTED** [review]
Nice - I will pull this down and test it out

File: ui/src/app/app-routing.module.ts (hunk 1 above)
Comment (ericsong) [reply ui/src/app/app-routing.module.ts:114]: I did try that actually and spent quite a bit of time trying to get it to work but the React `BrowserRouter` wouldn't update what it should render even though the URL would change.. Some of the things I tried were 
  - add the `withRouter` prop to all components
  - wrap in `<Switch>`
  - change the `NavStore.navigate*` functions to change routes by pushing to the History API which react-router uses as its backend

File: ui/src/app/components/app-router.tsx [hunk 2]
```diff
@@ -16,8 +16,18 @@ export const usePath = () => {
...
 
-export const withRouteData = WrappedComponent => ({routeData, ...props}) => {
-  routeDataStore.set(routeData);
+// TODO angular2react: I don't think this is really the right place to be making the store updates but it's the
+// best place I found while we're using both angular and react routers
+export const withRouteData = WrappedComponent => ({intermediaryRoute = false, routeData, ...props}) => {
+  const params = useParams();
+  console.log('using params');
+
+  if (!intermediaryRoute) {
```
Comment (calbach) [inline ui/src/app/components/app-router.tsx:25]: This should probably be inside a `useEffect()`, which reruns based on `params` changing. This render method's main body would otherwise run an arbitrary number of times, at non-deterministic timings.

In the future, I think urlParamsStore should be removed entirely in favor of `useParams`. For the route data, I'm not sure of the best pattern, though we should likely switch off this global state. Anyways - this seems good during the transition.

File: ui/src/app/components/copy-modal.tsx [hunk 3]
```diff
@@ -232,6 +232,8 @@ class CopyModalComponent extends React.Component<Props, State> {
   }
 
   goToDestinationWorkspace() {
+    currentWorkspaceStore.next(null);
```
Comment (calbach) [inline ui/src/app/components/copy-modal.tsx:235]: This seems suspect - shouldn't this be handled somewhere more centrally during navigation? It is probably impossible guard every scenario where the user switches between active workspaces, especially since they can use browser history to introduce unexpected transitions.

File: ui/src/app/components/help-sidebar.tsx [hunk 4]
```diff
@@ -397,37 +395,23 @@ export const HelpSidebar = fp.flow(
...
-            loadedLastSavedKey = true;
-          }
-        };
-      })();
-
-      this.loadLastSavedKey();
-      this.subscription = participantStore.subscribe(participant => this.setState({participant}));
-      this.subscription.add(setSidebarActiveIconStore.subscribe(activeIcon => {this.setActiveIcon(activeIcon); }));
-      this.subscription.add(routeDataStore.subscribe((newRoute, oldRoute) => {
+      // This is being set here instead of the constructor since I like showing the opening animation of the side
```
Comment (calbach) [inline ui/src/app/components/help-sidebar.tsx:402]: super nit: would generally avoid first-person in comments, e.g. "since I like showing" -> "in order to show", unless it's a temporary TODO comment.

File: ui/src/app/pages/access/access-renewal-page.tsx [hunk 5]
```diff
@@ -209,8 +208,7 @@ const RenewalCard = withStyle(renewalStyle.card)(
 
 
 // Page to render
-export const AccessRenewalPage = fp.flow(
-  withRouteData,
+export const AccessRenewal = fp.flow(
```
Comment (calbach) [inline ui/src/app/pages/access/access-renewal-page.tsx:211]: Why this change? If this is desired, I think the filename also needs to be updated.

File: ui/src/app/pages/app/component.ts [hunk 6]
```diff
@@ -93,14 +102,95 @@ export class AppComponent implements OnInit {
...
+        setSidebarActiveIconStore.next(null);
+      }
+    }));
+
+    this.subscriptions.push(urlParamsStore
+      .map(({ns, wsid}) => ({ns, wsid}))
+      .debounceTime(1000) // Kind of hacky but this prevents multiple update requests going out simultaneously
+      // due to urlParamsStore being updated multiple times while rendering a route.
+      // What we really want to subscribe to here is an event that triggers on navigation start or end
+      // Debounce 1000 (ms) will throttle the output events to once a second which should be OK for real life usage
```
Comment (calbach) [inline ui/src/app/pages/app/component.ts:130]: Famous last words..

Probably an important thing to note here is that the `recent resources` stuff is pretty much a best effort. Debouncing this by every second is totally fine, since we don't mind missing an occasional event.

File: ui/src/app/pages/app/component.ts [hunk 7]
```diff
@@ -177,6 +267,12 @@ export class AppComponent implements OnInit {
     });
   }
 
+  ngOnDestroy() {
+    for (const s of this.subscriptions) {
```
Comment (calbach) [inline ui/src/app/pages/app/component.ts:271]: nit: should probably be consistent with other file in terms of `for` loops vs `foreach` on unsubscribe

File: ui/src/app/pages/data/data-component.tsx [hunk 8]
```diff
@@ -78,202 +77,175 @@ const dataSetImg = '/assets/images/dataset-diagram.svg';
 
 const resourceTypesToFetch = [ResourceType.COHORT, ResourceType.COHORTREVIEW, ResourceType.CONCEPTSET, ResourceType.DATASET];
 
-export const DataComponent = withCurrentWorkspace()(class extends React.Component<
-  {workspace: WorkspaceData},
-  {activeTab: Tabs, resourceList: WorkspaceResource[], isLoading: boolean,
-    creatingConceptSet: boolean, existingDataSetName: string[], existingCohortName: string[],
-    existingCohortReviewName: string[], existingConceptSetName: string[]}> {
+export const DataComponent = withCurrentWorkspace()(({workspace}) => {
```
Comment (calbach) [inline ui/src/app/pages/data/data-component.tsx:80]: Arguably this change could be split out to reduce risk - unless this has some dependency on the other changes that I'm missing.

File: ui/src/app/pages/data/data-component.tsx [hunk 9]
```diff
@@ -78,202 +77,175 @@ const dataSetImg = '/assets/images/dataset-diagram.svg';
...
-      });
+      setIsLoading(true);
+      setResourceList(
+        (await workspacesApi().getWorkspaceResources(workspace.namespace, workspace.id, {typesToFetch: resourceTypesToFetch}))
+          .map(result => {
+            return {
+              ...result,
+            // TODO (RW-4682): Fix this nonsense
+              modifiedTime: result.modifiedTime ? new Date(Number(result.modifiedTime)).toDateString() : null
+            }; }));
```
Comment (calbach) [inline ui/src/app/pages/data/data-component.tsx:99]:
```suggestion
          .map(result => ({
            ...result,
            // TODO (RW-4682): Fix this nonsense
            modifiedTime: result.modifiedTime ? new Date(Number(result.modifiedTime)).toDateString() : null
          })));
```

File: ui/src/app/pages/profile/profile-page.tsx [hunk 10]
```diff
@@ -128,8 +127,7 @@ const getControlledTierContent = fp.flow(
   getCompleteOrBypassContent
 );
 
-export const ProfilePage = fp.flow(
-  withRouteData,
+export const ProfileComponent = fp.flow(
```
Comment (calbach) [inline ui/src/app/pages/profile/profile-page.tsx:130]: Same comment on file name

File: ui/src/app/pages/workspace/workspace-wrapper.tsx [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+  withCurrentWorkspace(),
+  withRouteConfigData()
+)(({workspace, routeConfigData}) => {
+  return <React.Fragment>
+    {workspace
+        ? <React.Fragment>
+          {!routeConfigData.minimizeChrome && <WorkspaceNavBarReact tabPath={routeConfigData.workspaceNavBarTab}/>}
+          <HelpSidebar pageKey={routeConfigData.pageKey}/>
+          <div style={{marginRight: '45px', height: !routeConfigData.contentFullHeightOverride ? 'auto' : '100%'}}>
+              <WorkspaceRoutes/>
```
Comment (calbach) [inline ui/src/app/pages/workspace/workspace-wrapper.tsx:19]: nit: we need the linter to start autofixing JSX, but pretty sure this should be 2sp indent right here

File: ui/src/app/utils/runtime-utils.tsx [hunk 12]
```diff
@@ -213,6 +213,10 @@ const useRuntime = (currentWorkspaceNamespace) => {
   // No cleanup is being handled at the moment.
   // When the user initiates a runtime change we want that change to take place even if they navigate away
   useEffect(() => {
+    if (!currentWorkspaceNamespace) {
+      return;
+    }
+
```
Comment (calbach) [inline ui/src/app/utils/runtime-utils.tsx:219]: I guess one thing to look at later (or we can try now), is whether the second param to `useEffect` is still correct here. Possibly that should be `currentWorkspaceNamespace`

File: ui/src/app/workspace-app-routing.tsx [hunk 13]
```diff
@@ -0,0 +1,228 @@
...
+const NotebookRedirectPage = withRouteData(NotebookRedirect);
+const ParticipantsTablePage = withRouteData(ParticipantsTable);
+const QueryReportPage = withRouteData(QueryReport);
+const WorkspaceAboutPage = withRouteData(WorkspaceAbout);
+const WorkspaceEditPage = withRouteData(WorkspaceEdit);
+
+// TODO angular2react: Adding memo here feels a little off but it was necessary to prevent workspace-wrapper from
+// rendering over and over again on page load, rendering (hah) the app unusable.
+// We should be able to refactor this once we are driving the entire app through React router.
+export const WorkspaceRoutes = React.memo(() => {
```
Comment (calbach) [inline ui/src/app/workspace-app-routing.tsx:42]: We should definitely refactor this ASAP, on the next step or after conversion. I think this function has too many caveats

**Review by calbach — COMMENTED** [review]
Nice work. All of the main logic looks reasonable to me, though I have concerns in a handful of places.

File: ui/src/app/app-routing.tsx [hunk 14]
```diff
@@ -243,170 +216,9 @@ export const AppRoutingComponent: React.FunctionComponent<RoutingProps> = ({onSi
               />}
           />
           <AppRoute
-              path='/workspaces/:ns/:wsid/about'
```
Comment (calbach) [inline ui/src/app/app-routing.tsx:246]: Breaking this up might be preferable, but if we'd prefer to keep it together another option is to pass in the inner routes as React children to the Workspace wrapper.

File: ui/src/app/components/copy-modal.tsx (hunk 3 above)
Comment (ericsong) [reply ui/src/app/components/copy-modal.tsx]: Hmm.. yeah there probably are a few other cases that are not covered. Although, I think this does cover a lot of the official UI flows since it's for all the copy to workspace workflows.

For some context on why this was added, it prevents the following bug.

1. User navigates somewhere
2. Components start rendering and fetching data but gets the workspace from the current store which hasn't been updated yet
3. Route event completes and current workspace store is updated based off of the url params info
4. Component rerender with new workspace fetched through the store

There's a brief window between steps 2 and 4 where the component shows data for the old workspace.

Comment (ericsong) [reply ui/src/app/components/copy-modal.tsx]: After typing this out, I realized that the delay in steps 2-4 is due to the get workspace call and not due to delay in the url params value being emitted. I added a `currentWorkspaceStore.next(null)` call to the top of that function to null out the current workspace before the GET request is made and it looks like that works too.

https://github.com/all-of-us/workbench/blob/6e699ae16b73d32b20d9199aa791e56a344a7996/ui/src/app/pages/app/component.ts#L158

File: ui/src/app/pages/access/access-renewal-page.tsx (hunk 5 above)
Comment (ericsong) [reply ui/src/app/pages/access/access-renewal-page.tsx:211]: Two React components had their `withRouteData` HOC embedded into the component itself instead of the app-routing component. This broke their tests because of how I'm using `withRouteData` to set store values and it's call to `useParams()` was breaking. We can switch this back later once we refactor `withRouteData` but I do think I prefer keeping the routing stuff completely separate from the React component.

The change in name was just to reflect that difference. I can change the filenames as well.

File: ui/src/app/pages/app/component.ts (hunk 6 above)
Comment (ericsong) [reply ui/src/app/pages/app/component.ts:130]: yeah.. I would have been more careful for something more important but this seemed fine here

File: ui/src/app/pages/data/data-component.tsx (hunk 8 above)
Comment (ericsong) [reply ui/src/app/pages/data/data-component.tsx:80]: There's a functional change here which fixes a bug in the previous implementation. Previously, changes to the workspace prop would rerender the component but _not_ reload new resources since that call is only made on `componentDidMount`.

Net result was that copying a resource to a different workspace would not show the resources for the new workspace, rather the old one. (Side note, I would have thought that angular would force the whole component to rerender but I guess that wasn't the case.)

I was working on the fix and thought that `useEffect()` was a good solution so I converted the component to hooks while I was here.

Ideally, I could have shipped this out first but ended up bundling it b/c of the pre vacation crunch.

File: ui/src/app/pages/workspace/workspace-wrapper.tsx (hunk 11 above)
Comment (ericsong) [reply ui/src/app/pages/workspace/workspace-wrapper.tsx]: yeah, agreed on adding the JSX linter

File: ui/src/app/utils/runtime-utils.tsx (hunk 12 above)
Comment (ericsong) [reply ui/src/app/utils/runtime-utils.tsx:219]: sure. I think throwing it in is safe. a new value for `currentWorkspaceNamespace` should trigger a new get runtime call.

Although, I don't think it matters at the moment since the help-sidebar is remounting on every page load and running this regardless.

File: ui/src/app/workspace-app-routing.tsx (hunk 13 above)
Comment (ericsong) [reply ui/src/app/workspace-app-routing.tsx:42]: agreed

**Comment by ericsong** [comment]
@calbach thanks for the review. Things look OK from my smoke test, waiting for puppeteer to see if any breaking changes were introduced.

File: ui/src/app/components/app-router.tsx [hunk 15]
```diff
@@ -16,8 +17,24 @@ export const usePath = () => {
...
+// best place I found while we're using both angular and react routers
+export const withRouteData = WrappedComponent => ({intermediaryRoute = false, routeData, ...props}) => {
+  const params = useParams();
+
+  useEffect(() => {
+    if (!intermediaryRoute) {
+      routeConfigDataStore.next(routeData);
+      routeDataStore.set(routeData);
+    }
+  }, [routeData]);
```
Comment (calbach) [inline ui/src/app/components/app-router.tsx:30]: nit: technically these should both also depend on intermediaryRoute - though understood that's static

**Review by calbach — APPROVED** [review]
I pulled down the original branch (not the latest commits) and played around with it a bit - I didn't find any issues. LGTM

**Comment by ericsong** [comment]
@calbach thanks for testing. I just did some more testing as well.

@als364 The test failure on puppeteer is a known flake so this PR should be good to merge. Up to you and @calbach what you guys want to do with this PR since I know it can cause a lot of merge conflicts.

**Comment by calbach** [comment]
I support merging to avoid too much rebasing when you return, but I will be out next week. THis is also critical path for further React conversion.

Per discussions today, the plan is to **not** do a release next week, given the short week and general outages, so there will at least be more people around the following week if something breaks with the release.

**Review by als364 — COMMENTED** [review]
I think CH captured all the comments I had wanted to make. Let's sync up at standup on order of merges.

**Review by als364 — APPROVED** [review]

**Comment by als364** [comment]
Puppeteer error is https://precisionmedicineinitiative.atlassian.net/browse/RW-6905. Proceeding with merge
